### PR TITLE
DDF-2528: Adding AttributeValidator for Location.COUNTRY_CODE

### PR DESCRIPTION
**Pull request by brjeter**
#### What does this PR do?
This PR adds an attribute validator to verify that the Location.COUNTRY_CODE attribute is a valid ISO 3166-1 alpha-3 country code.
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@glenhein @jrnorth @peterhuffer 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@kcwire 
@millerw8
#### How should this be tested?
- Run a full build.
- Verify all tests pass (specifically the ValidationParserSpecTest and CountryCodeValidatorTest)
- Test ingestion of a record with an invalid location.country-code:
  - Run local DDF instance.
  - Configure `Metacard Validation Filter Plugin` to not filter metacards with errors/warnings.
  - Configure `Catalog Federation Strategy` to show validation errors.
  - Ingest a record with an invalid `location.country-code` attribute value.
  - Search for the record and verify there is an error message under `Validation Errors` under the `Details` tab.
- Test ingestion of a record with a valid location.country-code:
  - See above, but verify there are no error messages under `Validation Errors` on the metacard.

#### Any background context you want to provide?
The Location.COUNTRY_CODE attribute is now populated by the GeoCoderPlugin. This validation has been added to verify the attribute is populated correctly. See https://github.com/codice/ddf/pull/1171

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-2528
#### Checklist:
- [x] Documentation Updated
	- [ ] Change log Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+     * {@inheritDoc}
+     * <p>
+     * Validates only the values of {@code attribute} that are {@link String}s.
+     */
+    @Override
+    public Optional<AttributeValidationReport> validate(Attribute attribute) {
+        Preconditions.checkArgument(attribute != null, "The attribute cannot be null.");
+
+        final String name = attribute.getName();
+        for (final Serializable value : attribute.getValues()) {
```
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: I like it. Fixed.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                countryCodesMap.forEach(report::addSuggestedValue);
+                return Optional.of(report);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private void setupISO3CountryCodes() {
+        countryCodesMap = new HashSet<String>();
```
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Fixed it.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java (hunk 1 above)
Comment (coyotesqrl) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: One additional thing you could do, to clean up that lambda in the middle, is extract a new private method and then invoke it with `this::newMethod`. You'd have to add an additional map call:
```java
return attribute.getValues().stream()
        .filter(String.class::isInstance)
        .map(String.class::cast)
        .filter(s -> !countryCodesMap.contains(s))
        .map(s -> name)
        .map(this::newMethod)
        .map(AttributeValidationReport.class::cast)
        .findFirst();
```

It has two advantages: firstly it makes the streaming code cleaner; secondly, it makes it more explicit that the value is being dropped on the floor after it's been determined it is not in the `countryCodesMap`.

Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: The trick is that the code in there requires the `name` variable for the error message, and I won't have access to that local variable in the private function. I'm not sure how to get around that other than setting the `name` as a field of the class which I'm not a huge fan of doing in this situation.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+     * <p>
+     * Validates only the values of {@code attribute} that are {@link String}s.
+     */
+    @Override
+    public Optional<AttributeValidationReport> validate(Attribute attribute) {
+        Preconditions.checkArgument(attribute != null, "The attribute cannot be null.");
+
+        final String name = attribute.getName();
+
+        return attribute.getValues().stream()
```
Comment (rzwiefel) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Don't think your formatter is right here. Might wanna check you have it installed and make sure this file is formatted.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java (hunk 1 above)
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Oops, I didn't catch that. But I was just realizing as I was changing the code that I actually need access to the `name` and the `String` value in order to print a more meaningful message in the report. Something like
```
report.addViolation(new ValidationViolationImpl(Collections.singleton(name),
                s + "is not a valid ISO_3166-1 Alpha3 country code.",
                ValidationViolation.Severity.ERROR));
```

Comment (coyotesqrl) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Cool. Then you don't want that map operation, definitely.

You still might consider extracting a method like `AttributeValidationReport newMethod(String name, String countryCode)` to make the code cleaner, but otherwise it's good with me.

Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Good feedback, thanks. I extracted the method out.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java (hunk 3 above)
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: The files have been formatted

File: catalog/core/catalog-core-validationparser/src/main/java/ddf/catalog/validation/impl/ValidationParser.java [hunk 4]
```diff
@@ -329,6 +330,9 @@ private AttributeValidator getValidator(Outer.Validator validator) {
             }
             return new RangeValidator(min, max);
         }
+        case "country": {
```
Comment (jrnorth) [inline catalog/core/catalog-core-validationparser/src/main/java/ddf/catalog/validation/impl/ValidationParser.java]: How about calling this `iso3_country` or something similar? We might want to add other types of country codes in the future.

File: catalog/core/catalog-core-validationparser/src/test/groovy/ddf/catalog/validation/impl/ValidationParserSpecTest.groovy [hunk 5]
```diff
@@ -131,7 +131,7 @@ class ValidationParserSpecTest extends Specification {
         attributeRegistry.lookup(attribute2Name).isPresent()
 
         and: "the two attribute validators are registered in the attribute validator registry"
```
Comment (jrnorth) [inline catalog/core/catalog-core-validationparser/src/test/groovy/ddf/catalog/validation/impl/ValidationParserSpecTest.groovy]: The messages for the assertions you changed need to be updated.

**Comment by jrnorth** [comment]
You'll need to update the JSON validator documentation to include the new country validator.

Also it looks like there's an empty file called "list" in this PR that should be removed.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    private void setupISO3CountryCodes() {
+        countryCodes = Arrays.stream(Locale.getISOCountries())
+                .map(iso2CC -> new Locale("", iso2CC))
+                .map(Locale::getISO3Country)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
```
Comment (millerw8) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: any reason we need to override equals/hashcode?

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+        final String attributeName = attribute.getName();
+
+        return attribute.getValues()
+                .stream()
+                .filter(String.class::isInstance)
+                .map(String.class::cast)
+                .filter(s -> !countryCodes.contains(s))
+                .map(s -> this.buildReport(attributeName, s))
+                .findFirst();
```
Comment (millerw8) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Does this mean if we are validating location.country-code which I believe is multi-valued, and we enter CSWD and WFDA as two values, we will only get warned on one of them?

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java (hunk 6 above)
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: I implemented them because other AttributeValidator implementations overrode them. I assumed the decision was made to do this for all validators for a particular reason. Perhaps to programmatically tell the difference between AttributeValidators without knowing the implementations.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java (hunk 7 above)
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Yes that's correct. Should it be behaving differently?

Comment (millerw8) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Yeah, I think aggregating all the invalid value warnings would be ideal.

**Comment by emmberk** [comment]
Successfully completed mvn clean install :+1:

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                countryCodeValue + " is not a valid ISO_3166-1 Alpha3 country code.",
+                ValidationViolation.Severity.ERROR));
+
+        //TODO: Should the country codes be included in the error report?
+        countryCodes.forEach(report::addSuggestedValue);
+        return report;
+    }
+
+    private void setupISO3CountryCodes() {
+        countryCodes = Arrays.stream(Locale.getISOCountries())
```
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Nope, that's good feedback. Were you thinking a static block to initialize the `countryCodes`?

Comment (rzwiefel) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: You don't even need a static block.
Could just change line 41 to be
```java
private static final Set<String> countryCodes = Arrays.stream(Locale.getISOCountries())
    .map(iso2CC -> new Locale("", iso2CC))
    .map(Locale::getISO3Country)
    .collect(Collectors.toSet());
```

Comment (rzwiefel) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/validation/impl/validator/CountryCodeValidator.java]: Unless you think a static block will look/read cleaner. Up to you

File: distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc [hunk 9]
```diff
@@ -1730,7 +1730,8 @@ The `validator` key must have a value of one of the following:
  * `arguments`: (unlimited) [list of strings: each argument is one case-sensitive, valid enumeration value]
   - `enumerationignorecase`
   * `arguments`: (unlimited) [list of strings: each argument is one case-insensitive, valid enumeration value]
-
+ - `iso3_country` (validates that a String is a valid ISO 3166-1 alpha-3 country code)
```
Comment (kcwire) [inline distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc]: It appears we have an ignoreCase validator but it is not installed by default or documented.

**Comment by ricklarsen** [comment]
Should this get a mention in the change log?

File: catalog/plugin/catalog-plugin-metacard-validation/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 10]
```diff
@@ -61,7 +61,7 @@
 
 
     <!-- Register in the OSGi Service Registry -->
-    <service ref="pre-ingest-plugin" interface="ddf.catalog.plugin.PreIngestPlugin"/>
+    <service ref="pre-ingest-plugin" interface="ddf.catalog.plugin.PreIngestPlugin" ranking="-1"/>
```
Comment (jrnorth) [inline catalog/plugin/catalog-plugin-metacard-validation/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Why was this changed to -1?

File: catalog/core/catalog-core-validationparser/src/test/groovy/ddf/catalog/validation/impl/ValidationParserSpecTest.groovy [hunk 11]
```diff
@@ -130,8 +130,8 @@ class ValidationParserSpecTest extends Specification {
         attributeRegistry.lookup(attribute1Name).isPresent()
         attributeRegistry.lookup(attribute2Name).isPresent()
 
-        and: "the two attribute validators are registered in the attribute validator registry"
-        attributeValidatorRegistry.getValidators(attribute1Name).size() == 2
+        and: "the three attribute validators are registered in the attribute validator registry"
```
Comment (jrnorth) [inline catalog/core/catalog-core-validationparser/src/test/groovy/ddf/catalog/validation/impl/ValidationParserSpecTest.groovy]: Need to update these descriptions.

File: catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/validation/impl/ISO3CountryCodeValidatorTest.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    @Test
+    public void testInvalidISO3CountryCodesIgnoreCase() {
+        validateWithErrors(new AttributeImpl("location.country_code",
+                Arrays.asList(new String[] {"zzz", "abc"})), 2, false);
+    }
+
+    private void validateNoErrors(final Attribute attribute, boolean ignoreCase) {
+        ISO3CountryCodeValidator validator = new ISO3CountryCodeValidator(ignoreCase);
+        final Optional<AttributeValidationReport> reportOptional = validator.validate(attribute);
+        assertThat(reportOptional.isPresent(), CoreMatchers.is(false));
```
Comment (jrnorth) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/validation/impl/ISO3CountryCodeValidatorTest.java]: You can just use the `is()` method you statically imported.

File: catalog/plugin/catalog-plugin-metacard-validation/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 10 above)
Comment (brjeter) [inline catalog/plugin/catalog-plugin-metacard-validation/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Talked with Keith about this. We wanted the `MetacardValidityMarkerPlugin` to be the last pre-ingest plugin to process just in case any pre-ingest plugins modify/add attributes to the metacard. In fact, part of the reason for this PR was because the GeoCoder plugin was added as a pre-ingest plugin that adds the `location.country-code` attribute to the metacard but this was happening after validation which defeated the purpose.

File: catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/validation/impl/ISO3CountryCodeValidatorTest.java (hunk 12 above)
Comment (brjeter) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/validation/impl/ISO3CountryCodeValidatorTest.java]: Fixed.
